### PR TITLE
Display unread messages count 

### PR DESCRIPTION
**Pull request by afmicc**
## Display unread messages count

Trello: [As a user I should be able to see the unread conversations count so I can easily identify them](https://trello.com/c/8sRrMSsl)

#### Description

* Add field `is_read` to the Message model
* Add a custom action to `MessageViewSet` which sets the messages as read. It receives the id of the last message read and sets all previous messages as read too.
* Refactor existing serializers to send the needed data to this action.
* Add a param to the endpoint which lists the conversations to indicate how many unread messages the user has.

### CODE REVIEW

File: applications/chats/serializers.py [hunk 1]
```diff
@@ -16,13 +16,20 @@ class Meta:
...
     def get_receiver(self, obj):
         current_user = self.context['request'].user
         receiver = obj.members.exclude(id=current_user.id).first()
         return UserSerializer(receiver).data
+
+    def get_unread_message_count(self, room):
+        current_user = self.context['request'].user
+        return (room.message_set.filter(is_read=False)
+                                .exclude(writer__id=current_user.id)
+                                .count())
```
Comment (brunomichetti) [inline applications/chats/serializers.py]:
```suggestion
        return (
            room.message_set.filter(
                is_read=False
            ).exclude(
                writer__id=current_user.id
            ).count()
        )
```

File: applications/chats/tests/test_messages_set_last_message_read.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+class MessagesSetLastMessageReadTest(APITestCase):
+    def setUp(self):
+        self.user = UserFactory(confirmed=True)
+        self.receiver = UserFactory(confirmed=True)
+        self.room = RoomFactory(members=[self.user, self.receiver])
+
+    def call_set_last_message_read(self, rooom, from_message):
+        url = reverse(
+            'message-set-last-message-read',
+            kwargs={'room': rooom.id, 'pk': from_message.id}
```
Comment (brunomichetti) [inline applications/chats/tests/test_messages_set_last_message_read.py]:
```suggestion
            kwargs={'room': rooom.id, 'pk': from_message.id},
```

File: applications/chats/tests/test_messages_set_last_message_read.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        )
+        return self.client.post(url)
+
+    def test_only_received_messages_read_given_message(self):
+        messages_count = 3
+        messages = MessageFactory.create_batch(
+            messages_count,
+            room=self.room,
+            writer=self.receiver
```
Comment (brunomichetti) [inline applications/chats/tests/test_messages_set_last_message_read.py]:
```suggestion
            writer=self.receiver,
```

File: applications/chats/tests/test_room_list.py [hunk 4]
```diff
@@ -44,13 +44,42 @@ def test_created_rooms_returns_room_and_receiver(self):
...
+            room=room,
+            writer=receiver,
+            is_read=True
+        )
+        MessageFactory.create_batch(
+            message_count,
+            room=room,
+            writer=receiver,
+            is_read=False
+        )
```
Comment (brunomichetti) [inline applications/chats/tests/test_room_list.py]:
```suggestion
        for value in [True, False]:
            MessageFactory.create_batch(
                message_count,
                room=room,
                writer=receiver,
                is_read=value
            )
```

File: applications/chats/views.py [hunk 5]
```diff
@@ -19,3 +21,13 @@ class MessageViewSet(mixins.ListModelMixin, viewsets.GenericViewSet):
     def get_queryset(self):
         room_id = self.kwargs['room']
         return self.request.user.room_set.get(id=room_id).message_set.all()
+
+    @action(detail=True, methods=['post'])
+    def set_last_message_read(self, request, **args):
+        message = self.get_object()
+        unread_messages = (self.get_queryset()
+                               .filter(is_read=False, id__lte=message.id)
+                               .exclude(writer__id=self.request.user.id))
```
Comment (brunomichetti) [inline applications/chats/views.py]:
```suggestion
        unread_messages = (
            self.get_queryset()
            .filter(is_read=False, id__lte=message.id)
            .exclude(writer__id=self.request.user.id)
       )
```

**Review by brunomichetti — APPROVED** [review]

File: applications/chats/serializers.py (hunk 1 above)
Comment (afmicc) [reply applications/chats/serializers.py]: I knew that it looked weird haha

File: applications/chats/serializers.py [hunk 6]
```diff
@@ -16,13 +16,24 @@ class Meta:
...
+
+    def get_unread_message_count(self, room):
+        current_user = self.context['request'].user
+        return (
+            room.message_set.filter(
+                is_read=False
+            ).exclude(
+                writer__id=current_user.id
+            ).count()
+        )
```
Comment (guilleijo) [inline applications/chats/serializers.py]: I think the recommended by black would be:
```suggestion
        return (
            room.message_set.filter(is_read=False)
            .exclude(writer__id=current_user.id)
            .count()
        )
```

File: applications/chats/serializers.py [hunk 7]
```diff
@@ -16,13 +16,24 @@ class Meta:
 class RoomSerializer(serializers.ModelSerializer):
     class Meta:
         model = Room
-        fields = ('id', 'created_at', 'topic', 'receiver', )
-        read_only_fields = ('id', 'created_at', 'topic', 'receiver', )
+        exclude = ('members', )
+        read_only_fields = ('id', 'created_at', 'topic', 'receiver', 'unread_message_count', )
```
Comment (guilleijo) [inline applications/chats/serializers.py]: No need for the last comma here. In exclude is necessary to indicate that is a tuple.
```suggestion
        read_only_fields = ('id', 'created_at', 'topic', 'receiver', 'unread_message_count')
```

File: applications/chats/tests/test_messages_set_last_message_read.py [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+        MessageFactory.create_batch(
+            2,
+            room=self.room,
+            writer=self.user
+        )
+        received_messages += MessageFactory.create_batch(
+            2,
+            room=self.room,
+            writer=self.receiver
+        )
```
Comment (guilleijo) [inline applications/chats/tests/test_messages_set_last_message_read.py]: Add trailing commas or use one line if it is less than 120 chars.

File: applications/chats/tests/test_messages_set_last_message_read.py [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+        db_messages = Message.objects.filter(writer__id=self.user.id)
+        self.assertFalse(db_messages[1].is_read)
+        self.assertFalse(db_messages[0].is_read)
+
+    def test_user_does_not_belong_to_room_respond_forbidden(self):
+        room = RoomFactory(members=[UserFactory(confirmed=True), self.receiver])
+        message_to_read = MessageFactory.create(
+            room=room,
+            writer=self.receiver
+        )
```
Comment (guilleijo) [inline applications/chats/tests/test_messages_set_last_message_read.py]: Same here.

File: applications/chats/tests/test_messages_set_last_message_read.py [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+        response = self.call_set_last_message_read(room, message_to_read)
+
+        self.assertEqual(response.status_code, status.HTTP_403_FORBIDDEN)
+
+    def test_message_does_not_belong_to_room_respond_not_found(self):
+        room = RoomFactory(members=[UserFactory(confirmed=True), self.receiver])
+        message_to_read = MessageFactory.create(
+            room=room,
+            writer=self.receiver
+        )
```
Comment (guilleijo) [inline applications/chats/tests/test_messages_set_last_message_read.py]: Same here.

File: applications/chats/tests/test_messages_set_last_message_read.py [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+        self.client.force_authenticate(user=self.user)
+        response = self.call_set_last_message_read(self.room, message_to_read)
+
+        self.assertEqual(response.status_code, status.HTTP_404_NOT_FOUND)
+
+    def test_no_logged_user_respond_unauthorized(self):
+        message_to_read = MessageFactory.create(
+            room=self.room,
+            writer=self.receiver
+        )
```
Comment (guilleijo) [inline applications/chats/tests/test_messages_set_last_message_read.py]: Same here.

File: applications/chats/tests/test_room_list.py [hunk 12]
```diff
@@ -44,13 +44,37 @@ def test_created_rooms_returns_room_and_receiver(self):
         self.assertEqual(len(response.data.get('results')), rooms_count)
         self.assertEqual(
             [key for key in response.json()['results'][0]],
-            ['id', 'created_at', 'topic', 'receiver', ]
+            ['id', 'topic', 'receiver', 'unread_message_count', 'created_at', ]
```
Comment (guilleijo) [inline applications/chats/tests/test_room_list.py]:
```suggestion
            ['id', 'topic', 'receiver', 'unread_message_count', 'created_at'],
```

File: applications/chats/tests/test_room_list.py [hunk 13]
```diff
@@ -44,13 +44,37 @@ def test_created_rooms_returns_room_and_receiver(self):
...
+    def test_created_room_returns_room_unread_message_count(self):
+        receiver = UserFactory(confirmed=True)
+        room = RoomFactory(members=[receiver, self.user])
+        message_count = 3
+        for value in [True, False]:
+            MessageFactory.create_batch(
+                message_count,
+                room=room,
+                writer=receiver,
+                is_read=value
```
Comment (guilleijo) [inline applications/chats/tests/test_room_list.py]:
```suggestion
                is_read=value,
```

File: applications/chats/tests/test_room_list.py [hunk 14]
```diff
@@ -44,13 +44,37 @@ def test_created_rooms_returns_room_and_receiver(self):
...
+        self.client.force_authenticate(user=self.user)
+        response = self.call_rooms_list()
+
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(response.data.get('count'), 1)
+        self.assertEqual(len(response.data.get('results')), 1)
+        room_responsed = response.json()['results'][0]
+        self.assertEqual(
+            room_responsed['unread_message_count'],
+            message_count
```
Comment (guilleijo) [inline applications/chats/tests/test_room_list.py]:
```suggestion
            message_count,
```

**Review by guilleijo — APPROVED** [review]
You could use black to format your files. We use 120 as line length so some of these could be done in one line, but sometimes is more readable using multiple lines. In those cases use trailing commas.

**Pull request merged**